Feature/domino

Ln11211 opened the pull request:
This PR closes issue [18 of LightAndShadow](https://github.com/Terasology/LightAndShadow/issues/18). Merging this PR solves the following tasks 

- [x] make dice texture a texture atlas
- [x] create a basic domino block template
- [x] create domino blocks (one for each possible value of a domino half)
- [x] create domino prefabs
- [x] Make domino fall momentarily when hit

Ln11211 committed:
Update items.png

1)Added dice faces to the items.png.
2)Added a blank dice face for domino.

Ln11211 committed:
Add dice faces to item.atlas

1) Added dice faces to items.atlas
     "one" represents the face with one dot and "two" with two dots and so on ...until six dots.
     "zero" represents the face with zero dots which will be used to make domino blocks.
2) Found a Typo in line 10, changed the "spaceTop" to "spadeTop" .

Ln11211 committed:
Fixed a typo

Found a Typo in line 10, changed the "spaceTop" to "spadeTop" .

Ln11211 committed:
Updated dice faces

As suggested, the dice face names have been changed. Added the prefix dice to the dice face names.

Ln11211 committed:
Created domino.block template

Created the domino block template. This block has all its properties derived from card.block template.

Ln11211 committed:
Created domino blocks

Created domino blocks (one for each possible value of a domino half).
There are 7 such domino blocks which are based on the template "domino" of LASR.

Ln11211 committed:
Added whiteDice assets to blockTiles

Added WhiteDice assets to blockTiles and also changed the individual domino block properties .

Ln11211 committed:
Fixed syntactical errors in domino blocks

Fixed JSON syntax in domino block files and named the blocks..
1) removed the trailing commas.
2) Added missing comma.

Ln11211 committed:
Added domino prefab

Added a domino prefab(Not sure if it works).

Ln11211 committed:
Added domino Prefabs

Added prefabs for remaining domino blocks.

Ln11211 committed:
Merge branch 'Terasology:develop' into feature/domino

Ln11211 committed:
Logic for the domino

Added the logic for a domino to fall when hit and rise when hit.

Ln11211 committed:
Merge branch 'Terasology:develop' into feature/domino

Ln11211 committed:
Tried Fixing an error

Tried fixing an error in line 131

Ln11211 commented on src/main/java/org/terasology/lightandshadowresources/logic/domino/DominoSystem.java:
```diff
@@ -0,0 +1,230 @@
+            Vector3fc dominoCenter = new Vector3f(bottomBlockPos).add(0, 0.5f, 0);
+            newDomino.addComponent(new LocationComponent(dominoCenter));
+
+            DominoComponent newDominoComp = new Domino.getComponent(DominoComponent.class);
```
@jdrueckert , Can you please take a look at this? should this be 
`DominoComponent newDominoComp = newDomino.getComponent(DominoComponent.class);` 
I tried that and there was no error but then the subsequent lines( from 132) showed `Cannot find symbol error` .
Can you help me in fixing this, please?

jdrueckert commented on src/main/java/org/terasology/lightandshadowresources/logic/domino/DominoSystem.java:
```diff
@@ -0,0 +1,230 @@
+            Vector3fc dominoCenter = new Vector3f(bottomBlockPos).add(0, 0.5f, 0);
+            newDomino.addComponent(new LocationComponent(dominoCenter));
+
+            DominoComponent newDominoComp = new Domino.getComponent(DominoComponent.class);
```
There is no `Domino` class, hence the compiler doesn't understand `new Domino`.
What exactly is your intention here? Do you want to create a new `DominoComponent`?
If so, you probably want to do `new DominoComponent` instead.

Ln11211 replied:
Yeah, I want to create a new DominoComponent and use it to make the domino fall and rise. To be honest, I copied the code from the door"s logic from CoreAdvancedaAssets and then modified it (refactored variables and some fields).

jdrueckert commented on src/main/java/org/terasology/lightandshadowresources/logic/domino/DominoSystem.java:
```diff
@@ -0,0 +1,230 @@
+            newDomino.addComponent(new LocationComponent(dominoCenter));
+
+            DominoComponent newDominoComp = new Domino.getComponent(DominoComponent.class);
+            newDominoComp.risenSide = risenSide;
```
Your `DominoComponent` does not have a field called `risenSide`. Did you forget to commit/push any local changes you made to the `DominoComponent` class? (same for `fallSide` and `isFallen`)

Ln11211 replied:
@jdrueckert Like I said I copied code from Door's component so I forgot to rename `openSide` and `closedSide` in `door.component`, my bad, I overlooked that part of the code. I will fix that right away.

Ln11211 committed:
More Fixes

1)Fixed some refactoring issues in DominoComponent.java
2)In the DominoSystem.java, fixed a component typo

Ln11211 committed:
Another Typo

In line 185 of DominoSystem.java

Ln11211 committed:
Removed Domino sounds

The removed code is
for rise
        if (domino.riseSound != null) {
            entity.send(new PlaySoundEvent(domino.riseSound, 1f));
        }

for fall
        if (domino.fallSound != null) {
            entity.send(new PlaySoundEvent(domino.fallSound, 1f));
        }

Ln11211 committed:
dominoRegion.prefab is added

This commits adds DominoRegion prefab

Ln11211 committed:
Updated dominoRegion.prefab

Just a small update

Ln11211 committed:
Removed dominoTop &DominoBottom

Removed DominoTop.block and DominoBottom.block block definitions as they are crashing the game.

jdrueckert commented:
Without having had time to dig too much into detail yet, there's one thing I noticed: In `DominoSystem` you're expecting the `DominoComponent` for both the `placeDomino` method (when receiving `ActivateEvent`) and the `onFrob` (btw what's "frob"?) method (also when receiving `ActivateEvent`).

However, as far as I've seen, you've never added a `DominoComponent` to a prefab, did you? - Don't get me wrong, from what we've discussed in the past, this is correct, because we cannot specify which exact domino top and bottom to use in the prefabs. However, this means, that you cannot expect the component when receiving an event on an entity, because that entity will not have that component. And if it does not have the component you're expecting, the method will not be executed.

I hope to find time for a more detailed test and review on Tuesday (my evening).

jdrueckert committed:
feat: support manual domino placement

- add domino item and item texture
- remove unnecessary "domino<Digit>" and "domino<Digit>Block" prefabs
- harden against DominoComponent not present in region prefab
- add DominoComponent to region prefab (not 100% sure we need this)
- add logic to randomly pick upper and lower domino halves when placing a domino

jdrueckert commented:
With this commit, you should be able to place dominoes manually by giving yourself domino items (`give domino` in in-game console) and placing them against a two-block high wall. The placed domino should have a randomly picked upper and lower half.

Try it out and look at the changes I did in the commit and please ask about anything that's unclear in how it works / why I did it that way.

I did not yet test / look into the domino interaction (making it fall, have it come up again after a timeout), but I hope to have time for it on Thursday.

jdrueckert commented:
So, had a look at the domino interaction:
1. dominos currently act like doors and swing open instead of falling over
2. when interacting with dominos, the tiles change to dominoZero - I guess this is due to my change setting the top and bottom half to dominoZero, this needs to be removed from the prefab and set correctly in DominoSystem to the randomly selected tiles (see my last commit).
